File: csharp/src/GuardianForge.Core/PolicyEvaluator.cs
using System.Text.RegularExpressions;

namespace GuardianForge.Core;

/// <summary>
/// The deterministic policy evaluator: it checks structured rules against events and emits
/// violations, returning the strictest mode and severity that fired. It runs first, before
/// any model - the fast deterministic path decides the clear cases.
/// </summary>
public sealed class PolicyEvaluator
{
    private readonly object _lock = new();
    private List<Policy> _policies;
    private readonly Dictionary<string, List<string>> _history = new();
    private Dictionary<string, Regex> _regexps = new();

    public PolicyEvaluator(IEnumerable<Policy> policies)
    {
        _policies = policies.ToList();
        Compile();
    }

    private void Compile()
    {
        _regexps = new Dictionary<string, Regex>();
        foreach (var p in _policies)
            foreach (var r in p.Rules)
                if (r.Op == Core.Op.Matches)
                    _regexps[r.RuleId] = new Regex(r.Value);
    }

    public List<Policy> Policies() { lock (_lock) return new List<Policy>(_policies); }

    public void SetPolicies(IEnumerable<Policy> policies)
    {
        lock (_lock)
        {
            _policies = policies.ToList();
            Compile();
        }
    }

    /// <summary>Evaluate an event, returning violations plus the strictest mode/severity.</summary>
    public (List<Violation> Violations, string Mode, string Severity) Evaluate(AgentEvent ev)
    {
        lock (_lock)
        {
            var fp = $"{ev.Type}:{ev.Tool}:{ev.Action}";
            if (!_history.TryGetValue(ev.AgentId, out var hist))
            {
                hist = new List<string>();
                _history[ev.AgentId] = hist;
            }
            hist.Add(fp);
            if (hist.Count > 256) hist.RemoveRange(0, hist.Count - 256);

            var violations = new List<Violation>();
            var mode = PolicyMode.Observe;
            var sev = Core.Severity.Info;
            foreach (var p in _policies)
            {
                foreach (var r in p.Rules)
                {
                    if (Matches(ev, r, hist))
                    {
                        violations.Add(new Violation
                        {
                            PolicyId = p.PolicyId, RuleId = r.RuleId, AgentId = ev.AgentId,
                            Severity = r.Severity, Reason = $"rule {r.RuleId} matched on {r.Field}",
                        });
                        if (PolicyMode.Rank(p.Mode) > PolicyMode.Rank(mode)) mode = p.Mode;
                        if (Core.Severity.Rank(r.Severity) > Core.Severity.Rank(sev)) sev = r.Severity;
                    }
                }
            }
            return (violations, mode, sev);
        }
    }

    private bool Matches(AgentEvent ev, Rule r, List<string> hist)
    {
        if (r.Op == Core.Op.CountOver)
        {
            var window = r.WindowSize > 0 ? r.WindowSize : hist.Count;
            var start = Math.Max(0, hist.Count - window);
            var count = 0;
            for (var i = start; i < hist.Count; i++)
                if (hist[i].Contains(r.Value, StringComparison.Ordinal)) count++;
            return count > r.Threshold;
        }

        var val = FieldValue(ev, r.Field);
        return r.Op switch
        {
            Core.Op.Equals => val == r.Value,
            Core.Op.NotEquals => val != r.Value,
            Core.Op.Contains => val.Contains(r.Value, StringComparison.Ordinal),
            Core.Op.Matches => _regexps.TryGetValue(r.RuleId, out var rx) && rx.IsMatch(val),
            _ => false,
        };
    }

    private static string FieldValue(AgentEvent ev, string field)
    {
        if (field == "type") return ev.Type;
        if (field == "tool") return ev.Tool;
        if (field == "action") return ev.Action;
        if (field.StartsWith("metadata.", StringComparison.Ordinal))
            return ev.Metadata.TryGetValue(field["metadata.".Length..], out var v) ? v : "";
        return "";
    }
}

File: java/src/main/java/com/guardianforge/PolicyEvaluator.java
package com.guardianforge;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * The deterministic policy evaluator.
 *
 * <p>Mirrors the Go {@code policy} package: it checks structured rules against events -
 * equality, inequality, contains, regexp, and windowed counts - and emits violations. It
 * runs before any LLM; only genuinely ambiguous cases are escalated to a model.
 */
public final class PolicyEvaluator {
    private List<Policy> policies;
    private final Map<String, List<String>> history = new HashMap<>();
    private final Map<String, Pattern> regexps = new HashMap<>();

    /**
     * Build an evaluator over a policy set, compiling {@code matches} rule regexes.
     *
     * @param policies the policy set
     */
    public PolicyEvaluator(List<Policy> policies) {
        this.policies = policies;
        compile();
    }

    private void compile() {
        regexps.clear();
        for (Policy p : policies) {
            for (Rule r : p.rules) {
                if (Op.MATCHES.equals(r.op)) {
                    try {
                        regexps.put(r.ruleId, Pattern.compile(r.value));
                    } catch (PatternSyntaxException e) {
                        // Skip rules whose regex fails to compile, like the Go reference.
                    }
                }
            }
        }
    }

    /**
     * Return a copy of the current policy set.
     *
     * @return the policies
     */
    public List<Policy> policies() {
        return new ArrayList<>(policies);
    }

    /**
     * Replace the policy set and recompile its regexes.
     *
     * @param p the new policy set
     */
    public void setPolicies(List<Policy> p) {
        this.policies = p;
        compile();
    }

    /**
     * Evaluate an event against every policy.
     *
     * @param ev the event to check
     * @return the violations plus the strictest mode and severity that fired
     */
    public PolicyResult evaluate(AgentEvent ev) {
        String fp = ev.type + ":" + ev.tool + ":" + ev.action;
        List<String> hist = history.getOrDefault(ev.agentId, new ArrayList<>());
        hist = new ArrayList<>(hist);
        hist.add(fp);
        if (hist.size() > 256) {
            hist = new ArrayList<>(hist.subList(hist.size() - 256, hist.size()));
        }
        history.put(ev.agentId, hist);

        List<Violation> violations = new ArrayList<>();
        String mode = PolicyMode.OBSERVE;
        String sev = Severity.INFO;
        for (Policy p : policies) {
            for (Rule r : p.rules) {
                if (matches(ev, r, hist)) {
                    violations.add(new Violation(p.policyId, r.ruleId, ev.agentId, r.severity,
                            "rule " + r.ruleId + " matched on " + r.field));
                    mode = PolicyMode.stricter(mode, p.mode);
                    sev = Severity.higher(sev, r.severity);
                }
            }
        }
        return new PolicyResult(violations, mode, sev);
    }

    private boolean matches(AgentEvent ev, Rule r, List<String> hist) {
        if (Op.COUNT_OVER.equals(r.op)) {
            int window = r.windowSize;
            if (window <= 0) {
                window = hist.size();
            }
            int start = 0;
            if (hist.size() > window) {
                start = hist.size() - window;
            }
            int count = 0;
            for (String h : hist.subList(start, hist.size())) {
                if (h.contains(r.value)) {
                    count++;
                }
            }
            return count > r.threshold;
        }

        String val = fieldValue(ev, r.field);
        switch (r.op) {
            case Op.EQUALS:
                return val.equals(r.value);
            case Op.NOT_EQUALS:
                return !val.equals(r.value);
            case Op.CONTAINS:
                return val.contains(r.value);
            case Op.MATCHES:
                Pattern rx = regexps.get(r.ruleId);
                return rx != null && rx.matcher(val).find();
            default:
                return false;
        }
    }

    private static String fieldValue(AgentEvent ev, String field) {
        if ("type".equals(field)) {
            return ev.type;
        }
        if ("tool".equals(field)) {
            return ev.tool;
        }
        if ("action".equals(field)) {
            return ev.action;
        }
        if (field.startsWith("metadata.")) {
            return ev.metadata.getOrDefault(field.substring("metadata.".length()), "");
        }
        return "";
    }
}
